### DISCUSSION TITLE
Current state of embedding 3D in Avalonia

### DISCUSSION THREAD

**Discussion by chris835** [AvaloniaUI/Avalonia#9393] [Q&A]
Hey fellas,

We're about 90% through our decision to move everything we have over to Avalonia, only remaining question we have is a properly supported way of embedding 3D graphics.

I've seen quite a few different PRs and discussions here about doing just that using D3D/OpenGL/Vulkan. What is the current state of this, and what is planned for the future?

Regards,
Chris

**Comment by niko-zvt** [comment]
@kekekeks Perhaps you can describe the current situation? I saw your rejected OpenTK [PR](https://github.com/AvaloniaUI/Avalonia/pull/6391).
https://github.com/AvaloniaUI/Avalonia/pull/6391#issuecomment-1063862226

**Reply by kekekeks** [reply]
OpenGL currently works across all platforms except for WebAssembly. On Windows you need to add `UseWgl` to platform options (note that we've recently fixed wglMakeCurrent errors).

You can find an example of using Silk.NET on desktop [here](https://github.com/kekekeks/Avalonia-Silk.NET-Example). Mobile platforms only support OpenGL ES, so the example needs to be modified accordingly.

There probably won't be an OpenTK integration package due to their love of static variables and assumptions that there is only one set of OpenGL function pointers in existence. And due to the fact that there are better OpenGL bindings like Silk.NET that don't require such integration package and just work out of the box.

Further plans:
- provide at least some way to host Vulkan content on Windows/Linux (planned this year)
- provide a way to consume DXGI share handle / IOSurface / external memory surface fd (no ETA)
- provide a way to use ANGLE-powered OpenGLES context on Windows, so the won't crash on graphics driver reset (no ETA)
